### PR TITLE
clean up usage of null/undefined

### PR DESCRIPTION
**Pull request by wjhsf**
This standardizes most of our methods to accept both null and undefined, and mostly return a value or undefined. Exceptions that I noticed are `canonicalDomain`, `getPublicSuffix`, `permuteDomain`, and `domainMatch`, which all return null. I don't remember why I skipped them when I originally did this, so I figured I'd just leave them for now. (I did most of this work a while ago, and evidently never pushed!)

This also introduces a helper type `Nullable<T>` for readability.

nb: #377 but without changes to cookie creation

### CODE REVIEW

File: lib/cookie/cookie.ts [hunk 1]
```diff
@@ -112,9 +112,19 @@ type ParseCookieOptions = {
...
+/**
+ * Parses a string into a Cookie object.
+ * @param str Cookie string to parse
+ * @returns `Cookie` object for valid string inputs, `undefined` for invalid string inputs,
+ * or `null` for non-string inputs or empty string
+ */
 function parse(
   str: string,
   options?: ParseCookieOptions,
+  // TBD: Should we change the API to have a single "invalid input" return type? I think `undefined`
```
[the remark this answers is not in the document]
Comment (colincasey) [reply lib/cookie/cookie.ts:124]: After browsing through this code some more, I'm leaning more towards we should do this now.

**Review by colincasey — CHANGES_REQUESTED** [review]

File: lib/cookie/cookie.ts [hunk 2]
```diff
@@ -387,18 +397,18 @@ const cookieDefaults = {
 type CreateCookieOptions = {
   key?: string
   value?: string
-  expires?: Date | 'Infinity' | null
+  expires?: Nullable<Date | 'Infinity'>
```
[the remark this answers is not in the document]
Comment (wjhsf) [reply lib/cookie/cookie.ts:401]: Yeah, the changes from the other PR will take precedence, since they make the types a bit tighter.

File: lib/memstore.ts [hunk 3]
```diff
@@ -54,20 +59,20 @@ export class MemoryCookieStore extends Store {
...
 
   override findCookie(
-    domain: string | null,
-    path: string | null,
-    key: string | undefined,
-  ): Promise<Cookie | undefined>
+    domain: Nullable<string>,
+    path: Nullable<string>,
+    key: Nullable<string>,
+  ): Promise<Nullable<Cookie>>
```
[the remark this answers is not in the document]
Comment (wjhsf) [reply lib/memstore.ts:65]: Yeah, I think `T | undefined` is a reasonable return type. There are a few helper functions like `canonicalDomain` that return `T | null`. Should we change those as well? They're top-level package exports, so it's _technically_ a breaking change, but they're not the things people actually use this package for.

[the remark this answers is not in the document]
Comment (colincasey) [reply lib/memstore.ts:65]: I think so. I only ported them over like that to maintain parity with the `v4` code but the return value inconsistencies have always bothered me 😅

**Review by colincasey — APPROVED** [review]

File: lib/cookie/cookie.ts [hunk 4]
```diff
@@ -276,6 +280,10 @@ function parse(
   return c
 }
 
+// TBD: `null` is valid JSON, but `undefined` is not. Options:
+// 1. Change this to return `undefined` - weird because it's not JSON
+// 2. Keep this as `null` - weird because it's a violation of our new convention
+// 3. Change *everything* to return `null` - a lot of work, maybe other edge cases that we can't change?
```
Comment (wjhsf) [inline lib/cookie/cookie.ts:286]: @colincasey thoughts on this?

File: test/parsing_test.js [hunk 5]
```diff
@@ -245,12 +241,12 @@ vows
         return Cookie.parse("\x08", true) || null;
       },
       "doesn't parse": function(c) {
-        assert.equal(c, null);
+        assert.equal(c, undefined);
```
Comment (colincasey) [inline test/parsing_test.js:244]: You missed the `Cookie.parse(...) || null` above which means the value of `c` here is actually `null` but the test still passes since `assert.equal` is used instead of `assert.strictEqual`.

```suggestion
        assert.equal(c, SHOULD_BE_UNDEFINED);
```

File: test/parsing_test.js [hunk 6]
```diff
@@ -533,10 +529,10 @@ vows
         };
       },
       "large one doesn't parse": function(c) {
-        assert.equal(c.cookie1, null);
+        assert.equal(c.cookie1, undefined);
```
Comment (colincasey) [inline test/parsing_test.js:532]: To be absolutely certain the value is `undefined`.
```suggestion
        assert.strictEqual(c.cookie1, undefined);
```

File: test/parsing_test.js [hunk 7]
```diff
@@ -533,10 +529,10 @@ vows
...
       },
       "small one doesn't parse": function(c) {
-        assert.equal(c.cookie2, null);
+        assert.equal(c.cookie2, undefined);
```
Comment (colincasey) [inline test/parsing_test.js:535]: To be absolutely certain the value is undefined.
```suggestion
        assert.strictEqual(c.cookie2, undefined);
```

File: test/parsing_test.js [hunk 8]
```diff
@@ -245,12 +241,12 @@ vows
         return Cookie.parse("\x08", true) || null;
```
Comment (wjhsf) [inline test/parsing_test.js:241]:
```suggestion
        return Cookie.parse("\x08", true) ?? SHOULD_BE_UNDEFINED;
```

**Review by colincasey — APPROVED** [review]
